### PR TITLE
Ensure queries are correctly cancelled when the user goes away.

### PR DESCRIPTION
**Pull request by tomwilkie**
Signed-off-by: Tom Wilkie <tom.wilkie@gmail.com>

### CODE REVIEW

File: pkg/querier/frontend/frontend.go [hunk 1]
```diff
@@ -250,33 +250,75 @@ func (f *Frontend) RoundTripGRPC(ctx context.Context, req *ProcessRequest) (*Pro
...
 			request.err <- err
 			return err
+		case <-originalCtx.Done():
+			return originalCtx.Err()
 		}
 
-		response, err := server.Recv()
-		if err != nil {
+		select {
+		case resp := <-recvChan:
```
Comment (gouthamve) [inline pkg/querier/frontend/frontend.go]: If we don't drain this the goroutine above will be stuck and we will have a goroutine leak.

Comment (gouthamve) [reply pkg/querier/frontend/frontend.go]: Yup infact both goroutines will be stuck if we don't drain the `errChan`

Comment (tomwilkie) [reply pkg/querier/frontend/frontend.go]: `errChan` is buffered so they wont leak that way; but I think they can leak as recvChan isn't drained.

**Comment by tomwilkie** [comment]
@gouthamve PTAL

**Pull request merged**